src/vm/runners/cairo_runner.rs

```diff
@@ -1135,13 +1136,84 @@ pub struct SegmentInfo {
     pub size: usize,
 }
 
+//* ----------------------
+//*   ExecutionResources
+//* ----------------------
+
 #[derive(Clone, Debug, Eq, PartialEq, Default)]
 pub struct ExecutionResources {
     pub n_steps: usize,
     pub n_memory_holes: usize,
     pub builtin_instance_counter: HashMap<String, usize>,
 }
 
+/// Returns a copy of the execution resources where all the builtins with a usage counter
+/// of 0 are omitted.
+impl ExecutionResources {
+    pub fn filter_unused_builtins(&self) -> ExecutionResources {
+        ExecutionResources {
+            n_steps: self.n_steps,
+            n_memory_holes: self.n_memory_holes,
+            builtin_instance_counter: self
+                .clone()
+                .builtin_instance_counter
+                .into_iter()
+                .filter(|builtin| !builtin.1.is_zero())
+                .collect(),
+        }
+    }
+}
+
+impl Add for ExecutionResources {
+    type Output = ExecutionResources;
+
+    fn add(self, _rhs: ExecutionResources) -> ExecutionResources {
+        let mut execution_resources_union: HashMap<String, usize> = HashMap::new();
+
+        self.builtin_instance_counter
+            .keys()
+            .filter(|k| _rhs.builtin_instance_counter.contains_key(*k))
+            .for_each(|k| {
+                execution_resources_union.insert(
+                    k.to_string(),
+                    self.builtin_instance_counter.get(k).unwrap()
+                        + _rhs.builtin_instance_counter.get(k).unwrap(),
+                );
+            });
+
+        return ExecutionResources {
+            n_steps: self.n_steps + _rhs.n_steps,
+            n_memory_holes: self.n_memory_holes + _rhs.n_memory_holes,
+            builtin_instance_counter: execution_resources_union,
+        };
+    }
+}
+
+impl Sub for ExecutionResources {
+    type Output = ExecutionResources;
+
+    fn sub(self, _rhs: ExecutionResources) -> ExecutionResources {
+        let mut execution_resources_union: HashMap<String, usize> = HashMap::new();
+
+        self.builtin_instance_counter
+            .keys()
+            .filter(|k| _rhs.builtin_instance_counter.contains_key(*k))
+            .for_each(|k| {
+                execution_resources_union.insert(
+                    k.to_string(),
+                    self.builtin_instance_counter.get(k).unwrap()
+                        - _rhs.builtin_instance_counter.get(k).unwrap(),
```

Comment:
This could cause a `SubstractionWithOverflow` error, could we use `saturating_sub` instead?